### ISSUE TITLE
HighlightActiveLine only lasts a few seconds

### ISSUE THREAD

**Issue by dracan** [RickStrahl/MarkdownMonster#501]
When HighlightActiveLine is active, the highlight disappears after a few seconds. It only reappears when you mouse-click a paragraph again. Then it disappears again a few seconds later.

**Comment by RickStrahl** [comment]
I don't see that but I suspect this depends on the theme you're using. Some of the themes - on purpose - have very subtle highlights. Looks like Twilight has literally no highlight, vsCode is super subtle and if you don't have good screen contrast you likely don't see it at all.

The syntax themes can be customized and you can copy and create new ones to match.

Try playing around with different syntax coloring themes - and what theme are you using?

**Comment by dracan** [comment]
This is just on the default theme. Can't remember the name, as I'm on my phone at the mo. It's not related to the subtlety, as I can see it clearly for a few seconds, then it dissappears.

**Comment by RickStrahl** [comment]
I can't duplicate that. I see that a few of the themes have very subtle highlights that are almost not visible (which is by design by the original editor themes we use courtesy of Ace Editor), but they don't change over time. The highlights are fixed.

So not sure what you're seeing that's different. 

Added a little stronger highlighting to the default `vscodedark` and `twilight` themes.  be6d0800e1896192c67f383185edcec0b4bbaabc

**Comment by dracan** [comment]
Here's an animated gif showing what I'm seeing. Notice how the highlight consistently disappears after 2 seconds. Also, I've just noticed that it doesn't appear at all for the very first paragraph.

![MarkdownMonster](https://user-images.githubusercontent.com/567988/54875812-71092480-4dfd-11e9-9951-00be99db3c76.gif)

**Comment by RickStrahl** [comment]
Oh that’s not line highlighting that’s vim or emacs  cursor highlighting. Which is a different beast altogether as it’s part of Ace Editors native vim/emacs behavior.

If you think this is important to you you can try raise this issue on the [ace editor repo](https://github.com/ajaxorg/ace/issues) and issues. 

Before you do though - take a look at the [kitchen sink example](https://ace.c9.io/build/kitchen-sink.html) on the Ace editor site and play around with different themes that they have and see if and how the behavior change so. From what little I know about  vim the way it works and displays in Ace is pretty standard...

**Comment by dracan** [comment]
Just to check we're talking about the same thing - are you looking at the blue paragraph highlight on the right-handle side (ie. preview pane) of my animated gif? This is what I'm talking about - not Vim's block cursor. If I set `KeyboardHandler` to _default_ instead of Vim - I still seee the blue highlight. Although, just checked, the blue highlight still appears when I disable `HighlightActiveLine` - so looks like it's not that either.

Basically, I just want to be able to quickly tell at a glance what paragraph in the preview pane I'm currently editing. For example, in vscode (which I normally use before trialling MM), it gives me a grey vertical line to indicate the current paragraph...

![image](https://user-images.githubusercontent.com/567988/54898850-ab3df900-4ec5-11e9-9eaa-b63e4af8698e.png)

The blue highlight seems what I'm after in MM - except it keeps on disappearing.

**Comment by RickStrahl** [comment]
Oh I'm sorry, I see what you mean now :smile:. You're referring to the highlight in the previewer not the editor. 

That's by design and is meant to only be a **visual hint** to where the last update was made. It updates only when you click the mouse in the editor, or you type and briefly pause - basically when the **document is refreshed in the preview**.

The main  reason for just flashing is that the highlight can and does interfere with some content on the page and we don't permanently compromise the preview. I've played around with a lot of combinations for this including the VS code like bar, but settled on the brief flashing visual hint that draws the eye if you are editing which is usually all you need to find your place.

**Comment by dracan** [comment]
Can this be made to also reappear when you move the cursor then?

**Comment by RickStrahl** [comment]
If you really want to do this you can do the following:

* Copy an existing Preview Theme
* In Theme.html add the following just above the `</head>` tag

```js
<script>
    function scrollToPragmaLine(lineno) {
  if (typeof lineno !== "number" || lineno < 0) return;
  setTimeout(function() {
      if (lineno < 3) {
        $("html").scrollTop(0);
        return;
      }

      try {
        var $el = $("#pragma-line-" + lineno);
        if ($el.length < 1) {
          var origLine = lineno;

          // try forward with 3 lines
          for (var i = 0; i < 3; i++) {
            lineno++;
            $el = $("#pragma-line-" + lineno);
            if ($el.length > 0)
              break;
          }
          // try backwards with 3 lines
          if ($el.length < 1) {
            lineno = origLine;
            for (var i = 0; i < 3; i++) {
              lineno--;
              $el = $("#pragma-line-" + lineno);
              if ($el.length > 0)
                break;
            }
          }
          if ($el.length < 1)
            return;
        }

        $el.addClass("line-highlight");
        
        // REMOVE THIS LINE
        //setTimeout(function() { $el.removeClass("line-highlight"); }, 1800);

        te.codeScrolled = new Date().getTime();
        if (lineno > 3)
          $("html").scrollTop($el.offset().top - 150);
      } catch (ex) {
      }
    },
    20);
}
</script>
```

**Comment by dracan** [comment]
Okay thanks - looks like I'll have to do the theme hack then. Shame it's not a setting though, as being able to know at a glance what paragraph you're editing feels like it should be native to a markdown editor.

But, that all sounds really negative - great work on MM - it's looking really good from my trial so far 😄

**Comment by dracan** [comment]
I've just tried the `Theme.html` edit, but that, unfortunately, doesn't help - as whilst the highlight isn't disappearing any more - it's only updating when I actually edit text or click with the mouse. So if I navigate with the keyboard, it doesn't update to where my cursor is.

**Comment by RickStrahl** [comment]
Ok so I added support for a configuration switch for `Editor.PreviewHighlightTimeout` that removes the need for the hack above. The value can be set in milliseconds or to 0 to leave the highlight on at all times.

1b053176cded9a21e938e732282868c15966b947

**Comment by RickStrahl** [comment]
Ok also implemented the highlighting for keyboard navigation. It's an easy fix, but we'll have to see how that works out for large documents where the refresh during scrolling might cause a problem.

These changes are in 1.16.3.6 and later. 23ab7eacc7c692f34247b9932a1eb932a31d46ea

**Comment by dracan** [comment]
Wow - that was quick! :) It's working much better now! Makes such a difference being able to see where I'm editing in the preview.

One last issue though, is that it doesn't seem to work when navigating via the Vim navigation keys - eg. hjkl.

Btw - I've just ended my trial of MM, and bought a copy! 😄 Thank you for your insanely quick replies and fixes 👍

**Comment by RickStrahl** [comment]
I think VIM mode takes over the keyboard in a different way that isn't even trapped. I'll take a look and see if there's a way around that, but I have feeling that VIM intercepts keystrokes before the editor gets control so that might not be doable.

**Comment by dracan** [comment]
Hi @RickStrahl. Does the closing of this issue mean there was no way to update the highlight when using the Vim navigation commands?

This changes made so far are making a _huge_ difference - so a massive thank you for this. Being able to effortlessly see what paragraph I'm currently working on at a glance in the preview window is such a huge UX improvement. I'm finding I'm staying in the zone much more because of this, as I'm not getting distracted each time I have to try and find my current paragraph in the preview window 😁👍

**Comment by RickStrahl** [comment]
Took another look and VIM navigation (at least for jk keys) should now work.

f3cec3fbd4c15946bc5595c86d996cfcdfa86c10

Here's more info on this issue I posted on StackOverflow only to figure it out a few minutes later :-)

https://stackoverflow.com/questions/55385599/how-to-detect-vim-navigation-mode-in-ace-editor/55386119#55386119

**Comment by dracan** [comment]
Nice work! Especially on posting your findings on Stackoverflow to help others with the same issue :) I hate it when you're searching for a problem to find someone has asked the question, solved it, but not posted their solution! Upvoted both your question and answer :)

Should this fix be in 1.16.4.2? I'm trying to download this, but whilst the download page says 1.16.4.2 - when I install it, the fix doesn't work, and the about box says I'm still using 1.16.4.2. I've downloaded it a couple of times to retry installing...

![image](https://user-images.githubusercontent.com/567988/55134753-9a89bf00-5121-11e9-834c-2a8ba112b87f.png)

![image](https://user-images.githubusercontent.com/567988/55134772-abd2cb80-5121-11e9-9110-1ab26ef01fff.png)

**Comment by dracan** [comment]
Also, the highlight doesn't seem to work at all for the first two paragraphs. Whether using Vim or not. This includes if I click with the mouse.
